compiler/semantic/op.go

```diff
@@ -934,3 +1007,50 @@ func (a *analyzer) semCallOp(call *ast.Call, seq dag.Seq) (dag.Seq, error) {
 	}
 	return append(seq, dag.NewFilter(c)), nil
 }
+
+// maybeConvertUserOp returns nil, nil if the call is determined to not be a
+// UserOp, otherwise it returns the compile op or the encountered error.
+func (a *analyzer) maybeConvertUserOp(call *ast.Call, seq dag.Seq) (dag.Op, error) {
+	op, err := a.scope.LookupOp(call.Name)
+	if op == nil || err != nil {
+		return nil, nil
+	}
+	if call.Where != nil {
+		return nil, fmt.Errorf("%s(): user defined operators cannot have a where clause", call.Name)
+	}
+	params, args := op.Params, call.Args
+	if len(params) != len(args) {
+		return nil, fmt.Errorf("%s(): %d arg%s provided when operator expects %d arg%s", call.Name, len(params), plural.Slice(params, "s"), len(args), plural.Slice(args, "s"))
+	}
+	exprs := make([]dag.Expr, len(op.Params))
+	for i, arg := range args {
+		var err error
+		exprs[i], err = a.semExpr(arg)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Add references so we can construct the graph of user op calls in order
+	// to later detect cycles.
+	if a.opDecl != nil {
+		a.opDecl.deps = append(a.opDecl.deps, op)
+	}
+	c := &dag.UserOpCall{
+		Kind:   "UserOpCall",
+		Name:   call.Name,
+		Exprs:  exprs,
+		Params: slices.Clone(params),
+	}
+	// The UserOp call needs a Seq but since UserOp's body may not be compiled
+	// defer this until the check has completed and we have determined there is
+	// not a cycle in op calls.
+	a.resolve = append(a.resolve, func() error {
+		if op.Body == nil {
+			// This should never happen so panic if it does.
+			panic("op.Body is nil")
+		}
+		c.Body = op.Body
+		return nil
+	})
```

Comment:
How about replacing UserOpCall's Body field with a `UserOp *dag.UserOp` field so this resolve step can go away?

Reply:
I actually thought about this but didn't consider this w.r.t not needing the resolve. Another advantage is we can drop the Params argument in UserOpCall. Let's try it out.